django/db/migrations/loader.py

```diff
@@ -17,6 +18,34 @@
 MIGRATIONS_MODULE_NAME = "migrations"
 
 
+def memoize_and_check_cycles(f):
+    """
+    Given a function with a single argument, memoize the results
+    while confirming that you don't end up in a cycle
+    """
+    # memo[value] = [done_calculating, result]
+    # This tracks in-progress calculations and raise a ValueError
+    # if a loop is detected
+    #
+    # (presence of the key indicates that we are calculating the result)
+    memo = {}
+
+    @wraps(f)
+    def wrapped_func(arg):
+        if arg in memo:
+            done_calculating, result = memo[arg]
+            if done_calculating:
+                return result
+            else:
+                raise ValueError(f"Cycle detected with {arg}")
+        memo[arg] = [False, None]
+        result = f(arg)
+        memo[arg] = [True, result]
+        return result
+
+    return wrapped_func
```

Comment:
I wrote this but now feel like it's a bit "engineered" (especially given that it's harder to write nice error messages this way), thinking now that I should just put in memoization and cycle tracking into the `has_been_applied` helper function directly